be/src/storage/lake/pk_index_loader.cpp

```diff
@@ -0,0 +1,200 @@
+            return;
+        }
+        OlapReaderStatistics stats;
+        seg_options.fs = *fs;
+        seg_options.stats = &stats;
+        seg_options.is_primary_keys = true;
+        seg_options.delvec_loader = std::make_shared<LakeDelvecLoader>(_tablet->update_mgr(), _builder);
+        seg_options.version = _version;
+        seg_options.tablet_id = _tablet->id();
+        seg_options.rowset_id = _rowset_id;
+        auto seg_itr = (*seg)->new_iterator(_schema, seg_options);
+        if (!seg_itr.ok()) {
+            st = seg_itr.status();
+            return;
+        }
+        while (true) {
+            auto chunk_info = std::make_shared<ChunkInfo>();
+            chunk_info->rowids.reserve(DEFAULT_CHUNK_SIZE);
+            chunk_info->chunk = ChunkHelper::new_chunk(_schema, DEFAULT_CHUNK_SIZE);
+            chunk_info->rssid = _rowset_id + _seg_id;
+            auto chunk = chunk_info->chunk.get();
+            st = (*seg_itr)->get_next(chunk, &chunk_info->rowids);
+            if (st.is_end_of_file()) {
+                break;
+            } else if (!st.ok()) {
+                return;
+            }
+            ExecEnv::GetInstance()->lake_pk_index_loader()->add_chunk(_tablet->id(), chunk_info);
+        }
+        (*seg_itr)->close();
+    }
+
+private:
+    Tablet* _tablet;
+    uint32_t _rowset_id;
+    uint32_t _seg_id;
+    const std::string _seg_name;
+    const Schema _schema;
+    int64_t _version;
+    const MetaFileBuilder* _builder;
+};
+
+PkIndexLoader::~PkIndexLoader() {
+    _contexts.clear();
+}
+
+Status PkIndexLoader::load(Tablet* tablet, const std::vector<RowsetPtr>& rowsets, const Schema& schema, int64_t version,
+                           const MetaFileBuilder* builder) {
+    uint64_t subtask_num = 0;
+    std::vector<std::shared_ptr<PkSegmentScanTask>> scan_tasks;
+    for (auto& rowset : rowsets) {
+        auto& rowset_meta = rowset->metadata();
+        for (uint32_t i = 0; i < rowset_meta.segments_size(); ++i) {
+            auto scan_task = std::make_shared<PkSegmentScanTask>(tablet, rowset->id(), i, rowset_meta.segments(i),
+                                                                 schema, version, builder);
+            bool ret = ExecEnv::GetInstance()->connector_scan_executor()->submit(
+                    workgroup::ScanTask(workgroup::WorkGroupManager::instance()->get_default_workgroup().get(),
+                                        [scan_task]() { scan_task->run(); }));
+            if (!ret) {
+                auto error_msg = fmt::format("Fail to submit pk segment scan subtask, tablet: {}", tablet->id());
+                LOG(ERROR) << error_msg;
+                return Status::IOError(error_msg);
+            }
+        }
+        subtask_num += rowset_meta.segments_size();
+    }
+
+    std::lock_guard<std::mutex> lg(_mutex);
+    auto context = std::make_shared<PkSegmentScanTaskContext>();
+    context->subtask_num = subtask_num;
+    _contexts.emplace(tablet->id(), context);
```

Comment:
what if the context already submitted?  need to check duplication of the context or not?

Reply:
done